src/wallet/wallet.cpp

```diff
@@ -690,6 +690,11 @@ bool CWallet::EncryptWallet(const SecureString& strWalletPassphrase)
         // bits of the unencrypted private key in slack space in the database file.
         dbw->Rewrite();
```

Comment:
In commit "After encrypting the wallet, reload the database environment"

Looking at the CDB::Rewrite implementation I noticed that it already has a while loop that sleeps until mapFileUseCount (for just one wallet) is 0. You could take advantage of this if you wanted by tweaking the while loop there to wait for all use counts in the environment to be 0 and then calling ReloadDbEnv inside CDB::Rewrite. This would be a simplification since it would let you get rid of the new condition variable, and it would also make the new waiting code more consistent with previous code.

This is just a suggestion, though. Your current implementation seems fine, too.

Reply:
I don't want to do anything to `CDB::Rewrite` since it is used in places other than `encryptwallet`.

Reply:
https://github.com/bitcoin/bitcoin/pull/12493#discussion_r169838060

> I don't want to do anything to CDB::Rewrite since it is used in places other than encryptwallet

I didn't realize Rewrite was called other places, and the current approach does seem fine. But if you did want to unify the UseCount waiting logic, I think you could do it in a pretty clean way by adding a `bool reload_env` or similar option to CDB::Rewrite. I think in terms code clarity, having the option would actually be an improvement over always reloading.

Reply:
I think I'll leave it as it is now.